sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1656,6 +1688,170 @@ pi_command_list_ptr_t _pi_queue::eventOpenCommandList(pi_event Event) {
+        ZE_CALL(zeCommandListAppendBarrier,
+                (CmdList->first, Event->ZeEvent, EventWaitList.Length,
+                 EventWaitList.ZeEventList));
+        return PI_SUCCESS;
+      };
+
+  // If we have a list of events to make the barrier from, then we can create a
+  // barrier on these and use the resulting event as our future barrier.
+  if (NumEventsInWaitList) {
+    // Get an arbitrary command-list in the queue.
+    pi_command_list_ptr_t CmdList;
+    if (auto Res = Context->getAvailableCommandList(this, CmdList, false))
+      return Res;
+
+    // Retain the events as they will be owned by the result event.
+    _pi_ze_event_list_t TmpWaitList;
+    if (auto Res = TmpWaitList.createAndRetainPiZeEventList(
+            NumEventsInWaitList, EventWaitList, this,
+            CmdList->second.isCopy(this)))
+      return Res;
+
+    // Insert the barrier into the command-list and execute.
+    if (auto Res = insertBarrierIntoCmdList(CmdList, TmpWaitList, *Event))
+      return Res;
+    if (auto Res = executeCommandList(CmdList))
+      return Res;
+
+    // Retain and save the resulting event for future commands.
+    PI_CALL(piEventRetain(*Event));
+    ActiveBarriers.push_back(*Event);
+    return PI_SUCCESS;
+  }
+
+  // Since there are no events to explicitly create a barrier for, we are
+  // inserting a queue-wide barrier. As such, the barrier will also encapsulate
+  // the active barriers, so we can release and clear the active barriers list.
+  // Doing it early prevents potential additional barriers from implicitly being
+  // appended.
+  for (pi_event &E : ActiveBarriers)
+    PI_CALL(piEventRelease(E));
+  ActiveBarriers.clear();
+
+  // Get command lists for each command queue.
+  std::vector<pi_command_list_ptr_t> CmdLists;
+  if (UseImmediateCommandLists) {
+    // If immediate command lists are being used, each will act as their own
+    // queue, so we must insert a barrier into each.
+    CmdLists.reserve(CommandListMap.size());
+    for (auto It = CommandListMap.begin(); It != CommandListMap.end(); ++It)
+      CmdLists.push_back(It);
+  } else if (ComputeQueueGroup.ZeQueues.empty() &&
+             CopyQueueGroup.ZeQueues.empty()) {
+    // If there are no queues, we get any available command list.
+    pi_command_list_ptr_t CmdList;
+    if (auto Res = Context->getAvailableCommandList(this, CmdList, false))
+      return Res;
+    CmdLists.push_back(CmdList);
+  } else {
+    // Get an available command list for all queues.
+    CmdLists.reserve(ComputeQueueGroup.ZeQueues.size() +
+                     CopyQueueGroup.ZeQueues.size());
+    for (auto QueueGroup : {ComputeQueueGroup, CopyQueueGroup}) {
+      bool UseCopyEngine = QueueGroup.Type != queue_type::Compute;
+      for (ze_command_queue_handle_t ZeQueue : QueueGroup.ZeQueues) {
+        pi_command_list_ptr_t CmdList;
+        if (auto Res = Context->getAvailableCommandList(
+                this, CmdList, UseCopyEngine, false, &ZeQueue))
```

Comment:
I don't fully understand the need for the new `ForcedCmdQueue` argument (this is the only place using it, right?)
Please explain and add source code comments clarifying it.

Reply:
Thanks, I think I understand it now and will take another look.
> This is simpler for immediate command-lists

Why is that? In theory, there can be multiple immediate command-lists for the same SYCL queue (attached to different CCS-s)

Reply:
> each act as their own command-queue so we just use each of them instead having to find a minimal subset that cover all command-queues

Not sure what you mean by that, but similarly with regular command-lists, when a SYCL queue needs a barrier we have to add it to all the active L0 immediate command-lists.

Reply:
> Not sure what you mean by that, but similarly with regular command-lists, when a SYCL queue needs a barrier we have to add it to all the active L0 immediate command-lists.

Exactly! But they are easier to find in this case because we need all of them, rather than just some of them like for the non-immediate variants.